tokio/tests/rt_metrics.rs

```diff
@@ -82,20 +82,52 @@ fn blocking_queue_depth() {
 }
 
 #[test]
-fn active_tasks_count() {
+fn num_active_tasks() {
     let rt = current_thread();
     let metrics = rt.metrics();
-    assert_eq!(0, metrics.active_tasks_count());
-    rt.spawn(async move {
-        assert_eq!(1, metrics.active_tasks_count());
-    });
+    assert_eq!(0, metrics.num_active_tasks());
+    rt.block_on(rt.spawn(async move {
+        assert_eq!(1, metrics.num_active_tasks());
+    }))
+    .unwrap();
+
+    assert_eq!(0, rt.metrics().num_active_tasks());
 
     let rt = threaded();
     let metrics = rt.metrics();
-    assert_eq!(0, metrics.active_tasks_count());
-    rt.spawn(async move {
-        assert_eq!(1, metrics.active_tasks_count());
-    });
+    assert_eq!(0, metrics.num_active_tasks());
+    rt.block_on(rt.spawn(async move {
+        assert_eq!(1, metrics.num_active_tasks());
+    }))
+    .unwrap();
+
+    // this is too flaky on multiple runtime threads due to relaxed orderings
+    // assert_eq!(0, metrics.num_active_tasks());
```

Comment:
It's not because the counter uses relaxed orderings. Rather, it's because we call `wake_join()` before `self.release()` here:

https://github.com/tokio-rs/tokio/blob/6fcd9c02176bf3cd570bc7de88edaa3b95ea480a/tokio/src/runtime/task/harness.rs#L324-L333

The decrement happens somewhere inside `self.release()`.

I'm happy to merge this without the assert, but please fix the comment.

Reply:
One thing we could do is to use a sleep-loop until the two values are equal. This way we assert that it goes back to zero eventually. You can fail the test if it's still non-zero after 10 seconds.

Reply:
Trying the sleep strategy now - I have updated the comment as well.